src/server/extractors.rs

```diff
@@ -0,0 +1,35 @@
+use regex::Regex;
+use actix_web::{
+    Error, FromRequest,
+};
+use futures::future::{ok, err, Ready};
+use actix_web::error::ErrorBadRequest;
+
+pub struct  ClientState {
+   extractedClientState: String,
+   validatedClientState: String,
+}
+
+impl FromRequest for ClientState {
+    type Config = ();
+    type Error = Error;
+    type Future = Ready<Result<Self, Self::Error>>;
+
+    fn from_request(req: &::actix_web::HttpRequest,
+                    payload: &mut actix_web::dev::Payload,) -> Self::Future {
+
+        let client_state: &str = req.headers().get("X-Client-State").unwrap()
+                                        .to_str().unwrap();
+
+        let re = Regex::new(r"\^[a-zA-Z0-9._-]{1,32}$").unwrap();
```

Comment:
we can wrap this with [`lazy_static!` ](https://docs.rs/regex/1.3.9/regex/#example-avoid-compiling-the-same-regex-in-a-loop)to prevent it from being recompiled when the function is re-entered.

Might also want to use something like
`r"^[\w\.\-]{1,32}$"`
`\w` is short hand for alpha-numeric and underscore. (think "`\w`ords") It can be more clear to do what you're doing and spell out exactly what's being matched, so whichever you find clearer works. 
`.` on it's own will match any character in regex. Possibly not what you want. Probably a good idea to escape it.
`-` can also be misread by some regex parsers as "between". A good idea to escape it as well.

Reply:
This RE was copied from the original Python code.

Careful: we actually don't want `\w` here as it matches all unicode alphanumeric characters. We just want the ascii versions.

Reply:
Ah, and today I learned something about rust regex.

Reply:
> we can wrap this with [`lazy_static!` ](https://docs.rs/regex/1.3.9/regex/#example-avoid-compiling-the-same-regex-in-a-loop)to prevent it from being recompiled when the function is re-entered.
> 
> Might also want to use something like
> `r"^[a-zA-Z0-9\.\-]{1,32}$"`

Did you forget about the underscore('_') or it should not be there?

> `\w` ~is short hand for alpha-numeric and underscore. (think \words") It can be more clear to do what you're doing and spell out exactly what's being matched, so whichever you find clearer works.~
> `.` on it's own will match any character in regex. Possibly not what you want. Probably a good idea to escape it.
> `-` can also be misread by some regex parsers as "between". A good idea to escape it as well.

Reply:
I did forget about the underscore. `\w` includes underscore, but as Phil noted, we don't want to use that. The underscore should not need to be escaped like `\.` and `\-` do.